pallets/elections/src/lib.rs

```diff
@@ -305,6 +463,8 @@ pub mod pallet {
         ///
         /// We calculate the supports for them for the sake of eras payouts.
         fn elect() -> Result<Supports<T::AccountId>, Self::Error> {
+            Self::kick_out_underperformed_non_reserved_validators();
```

Comment:
It would be better if we run it outside of `elect()` but with the guarantee that it is run before `elect()`. As you mentioned -- `elect()` currently does not mutate the state, would be great if it stayed that way...

Reply:
The reason it was implemented like this is that we need to do it once per era in a way that supports and targets calculated in `elect()` including kick outs. 

Reasons why it cannot be done in `start_session`:
* we cannot do it after `SessionManager::new_session` is called since this is just _after_ `elect()` is done. 
* we cannot do it before `SessionManager::new_session` since `pallet_staking::ErasStartSessionIndex` is not yet populated 
with the next era info - `SessionManager::new_session` does it
Then it gets more error-prone to get it done in other functions such as `end_session` or `start_session` since we're away (in time) from elections. Also performance-wise we're not bloating the election block as its complexity is `O(non_reserved_validators.len())`.

 With all that above, I think for this version it's fine to leave it as it is, not sure if spending more time is worth results.